stakedb/ticketpool.go

```diff
@@ -0,0 +1,326 @@
+// Copyright (c) 2018, The dcrdata developers.
+// See LICENSE for details.
+
+package stakedb
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/asdine/storm"
+	"github.com/decred/dcrd/chaincfg/chainhash"
+)
+
+// TicketPool contains the live ticket pool diffs (tickets in/out) between
+// adjacent block heights in a chain. Diffs are applied in sequence by inserting
+// and removing ticket hashes from a pool, represented as a map. A []PoolDiff
+// stores these diffs, with a cursor pointing to the next unapplied diff. An
+// on-disk database of diffs is maintained using the storm wrapper for boltdb.
+type TicketPool struct {
+	*sync.RWMutex
+	cursor int64
+	tip    int64
+	diffs  []PoolDiff
+	pool   map[chainhash.Hash]struct{}
+	diffDB *storm.DB
+}
+
+// PoolDiff represents the tickets going in and out of the live ticket pool from
+// one height to the next.
+type PoolDiff struct {
+	In  []chainhash.Hash
+	Out []chainhash.Hash
+}
+
+// PoolDiffDBItem is the type in the storm live ticket DB. The primary key (id)
+// is Height.
+type PoolDiffDBItem struct {
+	Height   int64 `storm:"id"`
+	PoolDiff `storm:"inline"`
+}
+
+// NewTicketPool constructs a TicketPool by opening the persistent diff db,
+// loading all known diffs, initializing the TicketPool values.
+func NewTicketPool(dbFile string) (*TicketPool, error) {
+	// Open ticket pool diffs database
+	db, err := storm.Open(dbFile)
+	if err != nil {
+		return nil, fmt.Errorf("failed storm.Open: %v", err)
+	}
+
+	// Load all diffs
+	var poolDiffs []PoolDiffDBItem
+	err = db.AllByIndex("Height", &poolDiffs)
+	if err != nil {
+		return nil, fmt.Errorf("failed (*storm.DB).AllByIndex: %v", err)
+	}
+	diffs := make([]PoolDiff, len(poolDiffs))
+	for i := range poolDiffs {
+		diffs[i] = poolDiffs[i].PoolDiff
+	}
+
+	// Construct TicketPool with loaded diffs and diff DB
+	return &TicketPool{
+		RWMutex: new(sync.RWMutex),
+		pool:    make(map[chainhash.Hash]struct{}),
+		diffs:   diffs,             // make([]PoolDiff, 0, 100000),
+		tip:     int64(len(diffs)), // number of blocks connected over genesis
+		diffDB:  db,
+	}, nil
+}
+
+// Close closes the persistent diff DB.
+func (tp *TicketPool) Close() error {
+	return tp.diffDB.Close()
+}
+
+// Tip returns the current length of the diffs slice.
+func (tp *TicketPool) Tip() int64 {
+	tp.RLock()
+	defer tp.RUnlock()
+	return tp.tip
+}
+
+// Cursor returns the current cursor, the location of the next unapplied diff.
+func (tp *TicketPool) Cursor() int64 {
+	tp.RLock()
+	defer tp.RUnlock()
+	return tp.cursor
+}
+
+// append grows the diffs slice and advances the tip height.
+func (tp *TicketPool) append(diff *PoolDiff) {
+	tp.tip++
+	tp.diffs = append(tp.diffs, *diff)
+}
+
+// trim is the non-thread-safe version of Trim.
+func (tp *TicketPool) trim() int64 {
+	if tp.tip == 0 || len(tp.diffs) == 0 {
+		return tp.tip
+	}
+	tp.tip--
```

Comment:
Is this the right spot? I ask because there is an error path underneath it.

Reply:
Yes, here's why.  The main function of `trim` is to drop the last element from `diffs`, which is essentially the opposite of `append` (so, `tip--` and `tp.diffs = tp.diffs[:len(tp.diffs)-1]`), but if the cursor would be invalidated, then we also have to retreat (apply the about-to-disappear diffs in reverse to the pool map).  Now the reason `retreatTo` comes before `tp.diffs[:len(tp.diffs)-1]` is because retreating may require the diff being chopped.  About `tip--`, `retreatTo` checks that the destination height is valid with `tip`.  So actually, the error should be a `panic` because this shouldn't happen and there's no way to recover aside from deleting the db and starting fresh.